subtests/docker_cli/namespaces/user_namespace.py

```diff
@@ -0,0 +1,138 @@
+r"""
+Summary
+---------
+
+The Linux kernel user namespace support provides additional security by
+enabling a process, and therefore a container, to have a unique range of
+user and group IDs which are outside the traditional user and group range
+utilized by the host system. Potentially the most important security
+improvement is that, by default, container processes running as the root
+user will have expected administrative privilege (with some restrictions)
+inside the container but will effectively be mapped to an
+unprivileged uid on the host.
+
+This test is only applicable on RHEL 7.4+ systems.
+
+Operational Summary
+----------------------
+
+#. Add the option in /etc/sysconfig/docker: --userns-remap=default
+#. Add subordinate uid and gid range in /etc/subuid and /etc/subgid:
+   dockremap:100000:65536
+#. Set system's max_user_namespaces to 1
+#. Restart docker service without any error
+#. Run an application within a container
+#. Check the user of application inside and outside the container:
+#. - The user inside the container should be root.
+#. - The user outside the container(on the host) should be the remapped user.
+
+"""
+
+
+import os
+import shutil
+from autotest.client import utils
+from dockertest import docker_daemon
+from dockertest.containers import DockerContainers
+from dockertest.dockercmd import DockerCmd
+from dockertest.images import DockerImages
+from dockertest.subtest import SubSubtest
+
+
+class user_namespace(SubSubtest):
+
+    def initialize(self):
+        super(user_namespace, self).initialize()
+        self.sub_stuff['restart_rst'] = None
+        self.sub_stuff['run_rst'] = None
+        self.sub_stuff['user_inside'] = ''
+        self.sub_stuff['user_outside'] = ''
+        self.sub_stuff['default_max'] = ''
+        with open('/proc/sys/user/max_user_namespaces') as f:
+            self.sub_stuff['max_user_ns'] = f.read().strip()
+
+        # Add the option in /etc/sysconfig/docker
+        if '--userns-remap=default' not in docker_daemon.cmdline():
+            docker_daemon.edit_options_file(add='--userns-remap=default')
+
+        # Add subordinate uid and gid range
+        self._change_sub_uid_gid('add')
+
+        # Set max_user_namespaces
+        self._change_max_userns('add')
+
+    def _change_sub_uid_gid(self, action):
+        if action == 'add':
+            if os.path.exists('/etc/subuid'):
+                shutil.copyfile('/etc/subuid', '/etc/subuid.bk')
```

Comment:
The standard way to do this is `os.rename`: that preserves mtime, mode, owner, and all file attributes. Then, when finishes, `os.rename` it back into place.